Mathlib/Combinatorics/SimpleGraph/Copy.lean

```diff
@@ -0,0 +1,203 @@
+-/
+import Mathlib.Combinatorics.SimpleGraph.Subgraph
+
+/-!
+# Copies of Simple Graphs
+
+This file introduces the concept of one simple graph containing a copy of another.
+
+## Main definitions
+
+* `SimpleGraph.Copy A B` is the type of copies of `A` in `B`, implemented as the subtype of
+  *injective* homomorphisms.
+
+* `SimpleGraph.IsIsoSubgraph` is the relation that `B` contains a copy of `A`, that
+  is, the type of copies of `A` in `B` is nonempty. This is equivalent to the existence of an
+  isomorphism from `A` to a subgraph of `B`.
+
+  This is similar to `SimpleGraph.IsSubgraph` except that the simple graphs here need not have the
+  same underlying vertex type.
+
+* `SimpleGraph.Free` is the predicate that `B` is `A`-free, that is, `B` does not contain a copy of
+  `A`. This is the negation of `SimpleGraph.IsIsoSubgraph` implemented for convenience.
+-/
+
+
+open Fintype
+
+namespace SimpleGraph
+
+variable {V α β γ : Type*} {G : SimpleGraph V}
+  {A : SimpleGraph α} {B : SimpleGraph β} {C : SimpleGraph γ}
+
+section Copy
+
+/-- The type of copies as a subtype of *injective* homomorphisms. -/
+abbrev Copy (A : SimpleGraph α) (B : SimpleGraph β) :=
+  { f : A →g B // Function.Injective f }
+
+/-- An injective homomorphism gives rise to a copy. -/
+abbrev Hom.toCopy (f : A →g B) (h : Function.Injective f) : Copy A B := ⟨f, h⟩
+
+/-- An embedding gives rise to a copy. -/
+abbrev Embedding.toCopy (f : A ↪g B) : Copy A B := Hom.toCopy f.toHom f.injective
+
+/-- An isomorphism gives rise to a copy. -/
+abbrev Iso.toCopy (f : A ≃g B) : Copy A B := Embedding.toCopy f.toEmbedding
+
+namespace Copy
+
+/-- A copy gives rise to a homomorphism. -/
+abbrev toHom : Copy A B → A →g B := Subtype.val
+
+@[simp] lemma coe_toHom (f : Copy A B) : ⇑f.toHom = f := rfl
+
+lemma injective : (f : Copy A B) → (Function.Injective f.toHom) := Subtype.prop
+
+instance : FunLike (Copy A B) α β where
+  coe f := DFunLike.coe f.toHom
+  coe_injective' _ _ h := Subtype.val_injective (DFunLike.coe_injective h)
+
+@[simp] lemma coe_toHom_apply (f : Copy A B) (a : α) : ⇑f.toHom a = f a := rfl
+
+/-- A copy induces an embedding of edge sets. -/
+def mapEdgeSet (f : Copy A B) : A.edgeSet ↪ B.edgeSet where
+  toFun := Hom.mapEdgeSet f.toHom
+  inj' := Hom.mapEdgeSet.injective f.toHom f.injective
+
+/-- A copy induces an embedding of neighbor sets. -/
+def mapNeighborSet (f : Copy A B) (a : α) :
+    A.neighborSet a ↪ B.neighborSet (f a) where
+  toFun v := ⟨f v, f.toHom.apply_mem_neighborSet v.prop⟩
+  inj' _ _ h := by
+    rw [Subtype.mk_eq_mk] at h ⊢
+    exact f.injective h
+
+/-- A copy gives rise to an embedding of vertex types. -/
+def asEmbedding (f : Copy A B) : α ↪ β := ⟨f, f.injective⟩
+
+/-- The identity copy from a simple graph to itself. -/
+@[refl] def refl (G : SimpleGraph V) : Copy G G := ⟨Hom.id, Function.injective_id⟩
+
+/-- The copy from a subgraph to the supergraph. -/
+def ofLE {G₁ G₂ : SimpleGraph V} (h : G₁ ≤ G₂) : Copy G₁ G₂ :=
```

Comment:
```suggestion
def ofLE (G₁ G₂ : SimpleGraph V) (h : G₁ ≤ G₂) : Copy G₁ G₂ :=
```
to avoid the knowledge of which graphs we're talking about being hidden inside an elided proof term